### PR TITLE
Allow dependency resolve rules and substitution rules to provide descriptive reason for selection

### PR DESCRIPTION
**Pull request by melix**
### Context

See #3719

Custom reasons are visible when running the dependency insight report:

substitution_360

### CODE REVIEW

**Comment by bigdaz** [comment]
@melix @jjohannes I didn't get around to reviewing this one.

A couple of quick comments:
- it would be great if we could avoid adding more inconsistency to the DSL. (like `withDescription()` vs `because`). But my DSL-design skills are below par.
- Seems like we missed adding support for `DependencySubstitution.useTarget()`

**Comment by melix** [comment]
FTR I used `because` only where the DSL was already not on par with what we have elsewhere, so that it reads better. So this is an _intentional_ inconsistency on a legacy DSL. I'm open to alternative names, but didn't find something that would fit the 2 contexts without being ambiguous or not well suited.

**Comment by bamboo** [comment]
I like `because` because it reads more like prose than `withReason` or `withDescription`. The inconsistency bothers me a bit too. My preference would be to rename `withDescription` to `because`and make `useVersion` and `useTarget` part of the builder interface so one could write something like:

```kotlin
resolutionStrategy.eachDependency {
    when (requested.name) {
       "foo" -> useVersion("2.0").because("I am in control")
       "bar" -> useTarget("org.test:bar:2.0").because("why not?")
       else  -> useVersion(requested.version)
    }
}
```

**Comment by melix** [comment]
@bamboo thanks for your input. There are 2 things to consider here:

1. we're trying to minimize the changes to an API that is going to be superceded by the component metadata rules we're implementing
2. we want to keep binary compatibility

Obviously 2. is stronger than 1, and that's the reason we can write:

`withDescription(...).useTarget(...)` but not `useTarget(...).withDescription(...)` (because changing the return type is a client breaking binary change).

However if you think that doesn't read too bad, I'm not against using:

`because(...).useTarget(...)`

**Comment by bamboo** [comment]
> because changing the return type is a client breaking binary change

Got it.

I like `because(...).useTarget(...)` better than `withDescription(...).useTarget(...)`.

File: subprojects/diagnostics/src/integTest/groovy/org/gradle/api/tasks/diagnostics/DependencyReportTaskIntegrationTest.groovy [hunk 1]
```diff
@@ -848,6 +849,55 @@ api (n)
 \\--- foo:foo:1.0 (n)
 
 (n) - Not resolved (configuration is not meant to be resolved)
+"""
+    }
+
+    @NotYetImplemented
```
Comment (jjohannes) [inline subprojects/diagnostics/src/integTest/groovy/org/gradle/api/tasks/diagnostics/DependencyReportTaskIntegrationTest.groovy]: :o: I would prefer to have this ` @ToBeImplemented` with an adjusted assertion. To not forget about this test. E.g. if we implement this with slightly different formatting, this test will still fail without us noticing.

Comment (melix) [reply subprojects/diagnostics/src/integTest/groovy/org/gradle/api/tasks/diagnostics/DependencyReportTaskIntegrationTest.groovy]: I have not been personally convinced that `@ToBeImplemented` is any better: https://gradle.slack.com/archives/C13AA53CK/p1505937273000344

**Comment by melix** [comment]
I have rebased the changes, moved from `withDescription` to `because` consistently, and added support for `DependencySubstitution.useTarget()`. It's ready for its second round of review!

**Pull request merged**